core/pkg/server/sender.go

```diff
@@ -967,75 +967,36 @@ func (s *Sender) sendFile(file *service.FilesItem) {
 	)
 	if err != nil {
 		err = fmt.Errorf("sender: sendFile: failed to get upload urls: %s", err)
-		s.logger.CaptureError("sender received error", err)
+		s.logger.CaptureError("sender: sendFile error", err)
 		return
 	}
-	headers := data.GetCreateRunFiles().GetUploadHeaders()
-	for _, f := range data.GetCreateRunFiles().GetFiles() {
-		fullPath := filepath.Join(s.settings.GetFilesDir().GetValue(), f.Name)
-		task := &filetransfer.Task{
-			Type:    filetransfer.UploadTask,
-			Path:    fullPath,
-			Name:    f.Name,
-			Url:     *f.UploadUrl,
-			Headers: headers,
-		}
 
-		task.SetProgressCallback(
-			func(processed, total int) {
-				if processed == 0 {
-					return
-				}
-				record := &service.Record{
-					RecordType: &service.Record_Request{
-						Request: &service.Request{
-							RequestType: &service.Request_FileTransferInfo{
-								FileTransferInfo: &service.FileTransferInfoRequest{
-									Type:      service.FileTransferInfoRequest_Upload,
-									Path:      fullPath,
-									Size:      int64(total),
-									Processed: int64(processed),
-								},
-							},
-						},
-					},
-				}
-				s.fwdChan <- record
-			},
+	if len(data.CreateRunFiles.Files) != 1 {
```

Comment:
is this correct? i think we can ask for signed url for multiple files, no?

Reply:
We only ever pass one path to CreateRunFiles() in this method; I'm assuming that means that there's only ever one path in the response. Is that not right?

The new version is safer too. In the old version, a man-in-the-middle attack could have been used to make the SDK upload an arbitrary file from the user's computer.

Reply:
in theory i think we could pass a batch of files and get a signed url for all these files, thus saving the multiple mutations. This would be a nice perf improvement, if we currently don't do it. Let's add a to-do to relax this and make it so we could do multiple files at once.

Reply:
the new `uploader.go` code I've been adding does do batches, at least during `run.finish()`. This code will be deleted from `sender.go` soon; I'm just trying to keep it correct in the meantime

Reply:
ah sweet! so you could also get batches of signed urls! very nice! 
we will get a good performance boost from this! super exciting :)